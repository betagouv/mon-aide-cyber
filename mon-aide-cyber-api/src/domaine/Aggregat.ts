import crypto from 'crypto';

export interface Aggregat {
  identifiant: crypto.UUID;
}

export class AggregatNonTrouve implements Error {
  message: string;
  name: string;

  constructor(typeAggregat: string) {
    this.name = '';
    this.message = `Le ${typeAggregat} demand√© n'existe pas.`;
  }
}
