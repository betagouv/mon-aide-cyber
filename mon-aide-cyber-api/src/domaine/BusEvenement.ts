import crypto from 'crypto';

export type Corps<T> = T;

export type Evenement<T> = {
  identifiant: crypto.UUID;
  type: TypeEvenement;
  date: Date;
  corps: Corps<T>;
};

export interface BusEvenement {
  publie<E extends Evenement<unknown>>(evenement: E): Promise<void>;
}

export interface ConsommateurEvenement {
  consomme<E extends Evenement<unknown>>(evenement: E): Promise<void>;
}

export type TypeEvenement =
  | 'DIAGNOSTIC_LANCE'
  | 'REPONSE_AJOUTEE'
  | 'RESTITUTION_LANCEE'
  | 'AIDANT_CREE'
  | 'AIDE_CREE'
  | 'DEMANDE_DEVENIR_AIDANT_CREEE'
  | 'MAIL_CREATION_COMPTE_AIDANT_ENVOYE'
  | 'MAIL_CREATION_COMPTE_AIDANT_NON_ENVOYE';
