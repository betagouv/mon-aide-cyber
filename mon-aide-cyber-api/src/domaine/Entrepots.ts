import { EntrepotDiagnostic } from '../diagnostic/Diagnostic';
import { EntrepotAidant } from '../authentification/Aidant';
import { EntrepotRestitution } from '../restitution/Restitution';
import { EntrepotAide } from '../aide/Aide';

import { EntrepotDemandeDevenirAidant } from '../gestion-demandes/devenir-aidant/DemandeDevenirAidant';
import { EntrepotStatistiques } from '../statistiques/statistiques';
import { EntrepotAnnuaireAidants } from '../annuaire-aidants/annuaireAidants';
import { EntrepotUtilisateur } from '../authentification/Utilisateur';

export interface Entrepots {
  diagnostic(): EntrepotDiagnostic;

  aidants(): EntrepotAidant;

  restitution(): EntrepotRestitution;

  aides(): EntrepotAide;

  demandesDevenirAidant(): EntrepotDemandeDevenirAidant;

  statistiques(): EntrepotStatistiques;

  annuaireAidants(): EntrepotAnnuaireAidants;

  utilisateurs(): EntrepotUtilisateur;
}
