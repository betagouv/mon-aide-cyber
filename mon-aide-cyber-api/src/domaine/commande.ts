export type Commande = {
  type: string;
};

export type Saga = Commande;

export interface CapteurSaga<S extends Saga, R> {
  execute(saga: S): Promise<R>;
}
export interface CapteurCommande<C extends Commande, R> {
  execute(commande: C): Promise<R>;
}

export interface BusCommande {
  publie<C extends Commande, R>(commande: C): Promise<R>;
}
