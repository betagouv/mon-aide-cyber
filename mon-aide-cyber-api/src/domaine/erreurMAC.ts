type Contexte =
  | 'Accès diagnostic'
  | 'Ajout réponse au diagnostic'
  | 'Termine le diagnostic'
  | 'Lance le diagnostic'
  | "Demande d'Authentification";

export class ErreurMAC implements Error {
  private constructor(
    public readonly contexte: string,
    public readonly erreurOriginelle: Error,
    public readonly message = '',
    public readonly name = '',
  ) {
    this.name = erreurOriginelle.name;
    this.message = erreurOriginelle.message;
  }

  public static cree(contexte: Contexte, erreur: Error) {
    return new ErreurMAC(contexte, erreur);
  }
}
