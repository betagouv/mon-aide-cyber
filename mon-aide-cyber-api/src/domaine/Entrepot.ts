import { Aggregat } from "./Aggregat";

export interface Entrepot<T extends Aggregat> {
  lis(identifiant: string): Promise<T>;

  persiste(entite: T): Promise<void>;

  typeAggregat(): string;
}
