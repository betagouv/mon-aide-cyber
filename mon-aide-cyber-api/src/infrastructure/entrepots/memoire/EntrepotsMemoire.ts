import { Entrepots } from '../../../domaine/Entrepots';

import {
  EntrepotAidantMemoire,
  EntrepotAideMemoire,
  EntrepotAnnuaireAidantsMemoire,
  EntrepotDemandeDiagnosticLibreAccesMemoire,
  EntrepotDemandeDevenirAidantMemoire,
  EntrepotDiagnosticMemoire,
  EntrepotProfilAidantMemoire,
  EntrepotRestitutionMemoire,
  EntrepotStatistiquesMemoire,
  EntrepotUtilisateurMemoire,
} from './EntrepotMemoire';
import { EntrepotDiagnostic } from '../../../diagnostic/Diagnostic';
import { EntrepotRestitution } from '../../../restitution/Restitution';
import { EntrepotAide } from '../../../aide/Aide';
import { EntrepotStatistiques } from '../../../statistiques/statistiques';
import { EntrepotAnnuaireAidants } from '../../../annuaire-aidants/annuaireAidants';
import { EntrepotUtilisateur } from '../../../authentification/Utilisateur';
import { EntrepotAidant } from '../../../espace-aidant/Aidant';
import { EntrepotProfilAidant } from '../../../espace-aidant/profil/profilAidant';
import { EntrepotDemandeDiagnosticLibreAcces } from '../../../diagnostic-libre-acces/CapteurSagaLanceDiagnosticLibreAcces';

export class EntrepotsMemoire implements Entrepots {
  private entrepotDiagnostic: EntrepotDiagnostic =
    new EntrepotDiagnosticMemoire();
  private entrepotAidants: EntrepotAidant = new EntrepotAidantMemoire();
  private entrepotRestitution: EntrepotRestitution =
    new EntrepotRestitutionMemoire();
  private entrepotAides: EntrepotAide = new EntrepotAideMemoire();
  private entrepotDemandeDevenirAidant =
    new EntrepotDemandeDevenirAidantMemoire();
  private entrepotStatistiques: EntrepotStatistiques =
    new EntrepotStatistiquesMemoire();
  private entrepotAnnuaireAidants: EntrepotAnnuaireAidants =
    new EntrepotAnnuaireAidantsMemoire();
  private entrepotUtilisateurs: EntrepotUtilisateur =
    new EntrepotUtilisateurMemoire();
  private entrepotProfilAidant: EntrepotProfilAidant =
    new EntrepotProfilAidantMemoire(
      this.entrepotAidants,
      this.entrepotUtilisateurs
    );
  private entrepotDemandesDiagnosticLibreAcces: EntrepotDemandeDiagnosticLibreAcces =
    new EntrepotDemandeDiagnosticLibreAccesMemoire();
  diagnostic(): EntrepotDiagnostic {
    return this.entrepotDiagnostic;
  }

  aidants() {
    return this.entrepotAidants;
  }

  restitution(): EntrepotRestitution {
    return this.entrepotRestitution;
  }

  aides() {
    return this.entrepotAides;
  }

  demandesDevenirAidant() {
    return this.entrepotDemandeDevenirAidant;
  }

  statistiques() {
    return this.entrepotStatistiques;
  }

  annuaireAidants() {
    return this.entrepotAnnuaireAidants;
  }

  utilisateurs(): EntrepotUtilisateur {
    return this.entrepotUtilisateurs;
  }

  profilAidant(): EntrepotProfilAidant {
    return this.entrepotProfilAidant;
  }

  demandesDiagnosticLibreAcces() {
    return this.entrepotDemandesDiagnosticLibreAcces;
  }
}
