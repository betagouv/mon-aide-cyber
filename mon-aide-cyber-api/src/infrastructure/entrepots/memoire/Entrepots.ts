import { Entrepots } from "../../../domaine/Entrepots";

import { EntrepotDiagnosticMemoire } from "./EntrepotsMemoire";
import { EntrepotDiagnostic } from "../../../diagnostic/Diagnostic";

export class EntrepotsMemoire implements Entrepots {
  private entrepotDiagnostic: EntrepotDiagnostic =
    new EntrepotDiagnosticMemoire();

  diagnostic(): EntrepotDiagnostic {
    return this.entrepotDiagnostic;
  }
}
