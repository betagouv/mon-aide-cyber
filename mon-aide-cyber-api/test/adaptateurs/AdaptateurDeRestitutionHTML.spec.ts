import { describe, it } from 'vitest';

import { unAdaptateurDeRestitutionHTML } from './ConstructeurAdaptateurRestitutionHTML';
import {
  AdaptateurDeRestitutionHTML,
  RestitutionHTML,
} from '../../src/adaptateurs/AdaptateurDeRestitutionHTML';
import { Entrepots } from '../../src/domaine/Entrepots';
import { EntrepotsMemoire } from '../../src/infrastructure/entrepots/memoire/EntrepotsMemoire';
import {
  desInformationsDeRestitution,
  uneRestitution,
} from '../constructeurs/constructeurRestitution';
import { uneMesurePriorisee } from '../constructeurs/constructeurMesure';
import { FournisseurHorlogeDeTest } from '../infrastructure/horloge/FournisseurHorlogeDeTest';
import crypto from 'crypto';

describe('Adapatateur de Restitution HTML', () => {
  const entrepots: Entrepots = new EntrepotsMemoire();

  describe('genère la restitution', () => {
    it('intègre les informations', async () => {
      const restitution = uneRestitution()
        .avecIdentifiant(crypto.randomUUID())
        .avecInformations(desInformationsDeRestitution().construis())
        .construis();
      entrepots.restitution().persiste(restitution);
      const adaptateurDeRestitutionHTML = unAdaptateurDeRestitutionHTML()
        .avecMesuresPrioritaires('une mesure prioritaire')
        .construis();

      const restitutionHTML =
        await adaptateurDeRestitutionHTML.genereRestitution(restitution);

      expect(restitutionHTML).toStrictEqual<RestitutionHTML>({
        autresMesures: '',
        indicateurs: '',
        informations: JSON.stringify(restitution.informations),
        mesuresPrioritaires: 'une mesure prioritaire',
      });
    });

    it('intègre des mesures prioritaires', async () => {
      const restitution = uneRestitution()
        .avecIdentifiant(crypto.randomUUID())
        .avecMesures([uneMesurePriorisee().construis()])
        .construis();
      entrepots.restitution().persiste(restitution);

      const adaptateurDeRestitutionHTML = unAdaptateurDeRestitutionHTML()
        .avecMesuresPrioritaires('une mesure prioritaire')
        .construis();

      const restitutionHTML =
        await adaptateurDeRestitutionHTML.genereRestitution(restitution);

      expect(restitutionHTML).toStrictEqual<RestitutionHTML>({
        autresMesures: '',
        indicateurs: '',
        informations: JSON.stringify(restitution.informations),
        mesuresPrioritaires: 'une mesure prioritaire',
      });
    });

    it('intègre un indicateur', async () => {
      const restitution = uneRestitution()
        .avecIdentifiant(crypto.randomUUID())
        .avecIndicateurs('thematique', 2)
        .construis();
      entrepots.restitution().persiste(restitution);

      const adaptateurDeRestitutionHTML = unAdaptateurDeRestitutionHTML()
        .avecIndicateurs("un graphe d'indicateurs")
        .construis();

      const restitutionHTML =
        await adaptateurDeRestitutionHTML.genereRestitution(restitution);

      expect(restitutionHTML).toStrictEqual<RestitutionHTML>({
        autresMesures: '',
        indicateurs: "un graphe d'indicateurs",
        informations: JSON.stringify(restitution.informations),
        mesuresPrioritaires: '',
      });
    });

    it('intègre les autres mesures', async () => {
      const adaptateurDeRestitutionHTML = unAdaptateurDeRestitutionHTML()
        .avecAutresMesures('une mesure non prioritaire')
        .construis();
      const mesurePriorisee = uneMesurePriorisee().construis();
      const restitution = uneRestitution()
        .avecIdentifiant(crypto.randomUUID())
        .avecMesures(Array(7).fill(mesurePriorisee))
        .construis();
      entrepots.restitution().persiste(restitution);

      const restitutionHTML =
        await adaptateurDeRestitutionHTML.genereRestitution(restitution);

      expect(restitutionHTML).toStrictEqual<RestitutionHTML>({
        autresMesures: 'une mesure non prioritaire',
        indicateurs: '',
        informations: JSON.stringify(restitution.informations),
        mesuresPrioritaires: '',
      });
    });
  });

  it('représente les informations du diagnostic', async () => {
    FournisseurHorlogeDeTest.initialise(
      new Date(Date.parse('2024-06-01T14:30:00+01:00')),
    );
    const restitution = uneRestitution()
      .avecIdentifiant('63a5fba6-efe0-4da1-ab6f-45d4cb330938')
      .avecIndicateurs('thematique', 0)
      .avecInformations(
        desInformationsDeRestitution()
          .avecSecteurActivite('')
          .avecZoneGeographique('')
          .construis(),
      )
      .avecMesures([])
      .construis();
    await entrepots.restitution().persiste(restitution);

    const informations = await new AdaptateurDeRestitutionHTML(
      new Map(),
    ).genereRestitution(restitution);

    expect(informations.informations).toMatchSnapshot();
  });
});
