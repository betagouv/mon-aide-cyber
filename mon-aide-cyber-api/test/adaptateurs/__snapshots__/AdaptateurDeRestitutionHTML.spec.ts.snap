// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Adapatateur de Restitution HTML > représente les informations du diagnostic 1`] = `"<h4>Informations</h4><div class="fr-mt-2w fr-mb-5w" id="information-contenu"><div class="fr-mb-2w" id="information-id">ID 63a5fba6-efe0-4da1-ab6f-45d4cb330938</div><div class="autres-informations"><p>Projet créé le<strong> 01.06.2024</strong> à<strong> 15:30</strong></p><p>Dernière Modification le<strong> 02.06.2024</strong> à<strong> 15:30</strong></p><p>Zone géographique :<strong> </strong></p><p>Secteur d'activité :<strong> </strong></p></div></div>"`;
